File: SW Expert Academy/D2/01954. 달팽이 숫자/달팽이 숫자.py
for tc in range(int(input())):
    N = int(input())

    print("#%d"%(tc+1))

    snail = [[0 for i in range(N)] for i in range(N)]

    i,j = 0,-1
    for k in range(N):
        j += 1
        snail[0][j] = k+1

    dir,num = 0,N+1
    for n in range(N*2-1, 1, -1):
        if dir == 0:
            for k in range(n//2):
                i += 1
                snail[i][j] = num
                num += 1
        elif dir == 1:
            for k in range(n//2):
                j -= 1
                snail[i][j] = num
                num += 1
        elif dir == 2:
            for k in range(n//2):
                i -= 1
                snail[i][j] = num
                num += 1
        else:
            for k in range(n//2):
                j += 1
                snail[i][j] = num
                num += 1

        dir = (dir+1)%4

    for s in snail:
        for si in s:
            print(si, end=' ')
        print()

File: SW Expert Academy/D2/01954. 달팽이 숫자/달팽이 숫자.cpp
#include <iostream>
#include <vector>
using namespace std;

int main() {
	int T; cin >> T;
	for (int tc = 0; tc < T; tc++) {
		int N; cin >> N;

		printf("#%d\n", tc + 1);

		vector<vector<int>> snail;
		for (int i = 0; i < N; i++) {
			vector<int> v(N);
			snail.push_back(v);
		}

		int i = 0, j = -1;
		for (int i = 0; i < N; snail[0][++j] = i++ + 1);

		int dir = 0;	// 0,1,2,3: 아래, 왼쪽, 위, 오른쪽
		int num = N + 1;
		for (N = N * 2 - 1; N > 1; N--) {
			if (dir == 0) for (int k = 0; k < N / 2; k++) snail[++i][j] = num++;
			else if (dir == 1) for (int k = 0; k < N / 2; k++) snail[i][--j] = num++;
			else if (dir == 2) for (int k = 0; k < N / 2; k++) snail[--i][j] = num++;
			else for (int k = 0; k < N / 2; k++) snail[i][++j] = num++;

			dir = (dir + 1) % 4;
		}

		// 출력
		for (int i = 0; i < snail.size(); i++) {
			for (int j = 0; j < snail.size(); j++)
				printf("%d ", snail[i][j]);
			cout << endl;
		}
	}
}
